File: examples/simple_webcam/java/client/SimpleWebcamClient.java
import com.robotraconteur.*;

import java.awt.image.*;
import java.io.*;

import javax.imageio.*;
import javax.swing.*;

import org.opencv.core.*;
import org.opencv.imgproc.*;
import org.opencv.imgcodecs.*;

import experimental.simplewebcam3.*;

// Simple client to read images from a Webcam server
// and display the images

public class SimpleWebcamClient
{

    /**
     * @param args
     */
    public static void main(String[] args)
    {

        String url = "rr+tcp://localhost:22355?service=multiwebcam";
        if (args.length > 0)
        {
            url = args[0];
        }

        // Load the opencv library
        nu.pattern.OpenCV.loadShared();

        // Use ClientNodeSetup to initialize node
        ClientNodeSetup setup = new ClientNodeSetup();
        try
        {
            RobotRaconteurNode.s().registerServiceType(new experimental__simplewebcam3Factory());

            // Connect to service
            WebcamHost c_host = (WebcamHost)RobotRaconteurNode.s().connectService(
                url, null, null, null, "experimental.simplewebcam3.WebcamHost");

            // Get the Webcam objects from the "Webcams" objref
            Webcam c1 = c_host.get_webcams(0);
            Webcam c2 = c_host.get_webcams(1);

            // Capture an image and convert to OpenCV image type
            Mat frame1 = WebcamImageToMat(c1.capture_frame());
            Mat frame2 = WebcamImageToMat(c2.capture_frame());

            // Show the images
            showImage(c1.get_name(), frame1);
            showImage(c2.get_name(), frame2);
        }
        finally
        {
            setup.finalize();
        }
    }

    // Convert WebcamImage to OpenCV format
    public static Mat WebcamImageToMat(WebcamImage image)
    {
        Mat mat = new Mat(image.height, image.width, CvType.CV_8UC3);
        mat.put(0, 0, image.data.value);
        return mat;
    }

    public static void showImage(String title, Mat img)
    {
        MatOfByte matOfByte = new MatOfByte();

        Imgcodecs.imencode(".bmp", img, matOfByte);
        byte[] byteArray = matOfByte.toArray();
        BufferedImage bufImage = null;
        try
        {
            InputStream in = new ByteArrayInputStream(byteArray);
            bufImage = ImageIO.read(in);
            JFrame frame = new JFrame();
            frame.setTitle(title);
            frame.getContentPane().add(new JLabel(new ImageIcon(bufImage)));
            frame.setDefaultCloseOperation(JFrame.DISPOSE_ON_CLOSE);
            frame.pack();
            frame.setVisible(true);
        }
        catch (Exception e)
        {
            e.printStackTrace();
        }
    }
}

File: examples/simple_webcam/cpp/client/simple_webcam_client.cpp

#include <RobotRaconteur.h>
#include "robotraconteur_generated.h"

#include <opencv2/highgui/highgui.hpp>

// Do not use aliases in header files!
namespace RR = RobotRaconteur;
namespace cam = ::experimental::simplewebcam3;

// Simple client to read images from a Webcam server
// and display the images

// This exmaple expects the simple_webcam_service_multi multi camera service to be running

// Convert WebcamImage to OpenCV format
cv::Mat webcam_image_to_mat(cam::WebcamImagePtr image)
{
    cv::Mat frame2(image->height, image->width, CV_8UC3);
    memcpy(frame2.data, &image->data->at(0), image->data->size());
    return frame2;
}

int main(int argc, char* argv[])
{
    std::string url = "rr+tcp://localhost:22355?service=multiwebcam";
    if (argc > 1)
    {
        url = argv[1];
    }

    try
    {
        // Use node setup to help initialize client node
        RR::ClientNodeSetup node_setup(ROBOTRACONTEUR_SERVICE_TYPES);

        cam::WebcamHostPtr c_host = RR::rr_cast<cam::WebcamHost>(RR::RobotRaconteurNode::s()->ConnectService(
            url, "", nullptr, NULL, "experimental.simplewebcam3.WebcamHost"));

        // Get the Webcam objects from the "Webcams" objref
        cam::WebcamPtr c1 = c_host->get_webcams(0);
        cam::WebcamPtr c2 = c_host->get_webcams(1);

        // Capture an image and convert to OpenCV image type
        cv::Mat frame1 = webcam_image_to_mat(c1->capture_frame());
        cv::Mat frame2 = webcam_image_to_mat(c2->capture_frame());

        // Show image
        cv::imshow("Left", frame1);
        cv::imshow("Right", frame2);

        // Wait for enter key to be pressed
        cv::waitKey();

        // Close the image viewers
        cv::destroyAllWindows();

        return 0;
    }
    catch (std::exception& e)
    {
        std::cout << "Error occured in client: " << std::string(e.what()) << std::endl;
        return 1;
    }
}
